src/title.js

```diff
@@ -0,0 +1,437 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module heading/title
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+import Paragraph from '@ckeditor/ckeditor5-paragraph/src/paragraph';
+
+import ViewWriter from '@ckeditor/ckeditor5-engine/src/view/downcastwriter';
+import {
+	needsPlaceholder,
+	showPlaceholder,
+	hidePlaceholder,
+	enablePlaceholder
+} from '@ckeditor/ckeditor5-engine/src/view/placeholder';
+
+// List of model elements that are allowed to be renamed to title element.
+const allowedToBeTitle = new Set( [ 'heading1', 'heading2', 'heading3', 'paragraph' ] );
+
+/**
+ * The Title plugin.
+ *
+ * It splits the document into `Title` and `Body` sections.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class Title extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'Title';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ Paragraph ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const model = editor.model;
+
+		/**
+		 * Reference to the empty paragraph in the body
+		 * created when there's no element in the body for the placeholder purpose.
+		 *
+		 * @private
+		 * @type {null|module:engine/model/element~Element}
+		 */
+		this._bodyPlaceholder = null;
+
+		// To use Schema for disabling some features when selection is inside the title element
+		// it's needed to create the following structure:
+		//
+		// <title>
+		//     <title-content>The title text</title-content>
+		// </title>
+		//
+		// See: https://github.com/ckeditor/ckeditor5/issues/2005.
+		model.schema.register( 'title', { inheritAllFrom: '$block' } );
+		model.schema.register( 'title-content', {
+			inheritAllFrom: '$block',
+			allowIn: 'title'
+		} );
+
+		model.schema.addChildCheck( ( context, childDefinition ) => {
+			// Allow `title` element only directly inside a root.
+			if ( childDefinition.name === 'title' && !context.endsWith( '$root' ) ) {
+				return false;
+			}
+
+			// Allow only `title-content` inside `title`.
+			if (
+				context.endsWith( 'title' ) && childDefinition.name !== 'title-content' ||
+				childDefinition.name === 'title-content' && !context.endsWith( 'title' )
```

Comment:
Ugh.. I was wrong because of `inheritAllFrom` there. So Now I'm not sure if we shouldn't make this more specific in the first place? Either by inheriting only those required options or by manually defining the `title` and `title-content` with full schema.

It looks like more straight-forward then allowing all there and then adding callback to check only those allowed scenarios.

Reply:
You are right. I'll define `title` and `title-content` manually.